### PR TITLE
connect_with_proxy

### PR DESCRIPTION
**Pull request by niuhuan**
#126  

Add proxy field to InitParms. Connect server with proxy if set.

### CODE REVIEW

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [inline lib/grammers-mtsender/src/lib.rs:9]: This should not be removed.

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [inline lib/grammers-mtsender/src/lib.rs:79]: This is an unfortunate (tiny) performance hit for the use-case where no proxies are used, but that can be solved with a new generic parameter.

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [inline lib/grammers-mtsender/src/lib.rs:142]: Why was this changed from the generic parameter?

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [inline lib/grammers-mtsender/src/lib.rs:146]: This will need to change before I can merge it. I propose we have a separate method, called something like `connect_with_proxy`, which is just like `connect`, but also accepts the proxy URL.

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [inline lib/grammers-mtsender/src/lib.rs:567]: Why was the generic removed? (Same as before.)

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [inline lib/grammers-mtsender/src/lib.rs:614]: Why was the generic removed? (Same as before.)

**Review by Lonami — CHANGES_REQUESTED** [review]
Thanks for the PR. If this solves your issues, I'm happy to merge, once we resolve the problem around using the environment to decide whether or not to use a proxy, and the other small nits.

**Comment by Lonami** [comment]
Note that while I'm OK with this code being in the main repository I still have no plans to directly support it (that will probably still be on the community). I just have no personal interest in dealing with proxies.

**Comment by niuhuan** [comment]
Thanks code review, this taught me a lot of knowledge , I will try to modify it according to your ideas.

File: lib/grammers-mtsender/src/lib.rs
Comment (niuhuan) [reply lib/grammers-mtsender/src/lib.rs:614]: Sorry, I'not found way convert ToSocketAddress to ToTargetAddress, I want it to run first

**Comment by niuhuan** [comment]
Had modify to `bot_sign_in_with_proxy`,  `request_login_code_with_proxy`, `connect_with_proxy`

File: lib/grammers-mtsender/src/lib.rs
Comment (niuhuan) [reply lib/grammers-mtsender/src/lib.rs:79]: Do you have any suggestions here?

**Comment by niuhuan** [comment]
I think must be known type befor use generic .

And this really bothered me. 

捕获

**Comment by niuhuan** [comment]
If anyone has spare time, I hope to solve this problem ( Make the code look better ). Proxy is too important for Chinese Mainland.

File: lib/grammers-mtsender/src/lib.rs
Comment (Lonami) [reply lib/grammers-mtsender/src/lib.rs:79]: No, it's fine to leave this as-is for now.

**Review by Lonami — CHANGES_REQUESTED** [review]
I hadn't thought about this before, but after seeing the changes to add `bot_sign_in_with_proxy` and `request_login_code_with_proxy`, I am confident that the right solution would be to add:

```rust
proxy_url: Option<String>,
```

to [`InitParams`](https://github.com/Lonami/grammers/blob/ad3fd4b7facafccc1eeacbf1df11a67b58d52e60/lib/grammers-client/src/client/client.rs#L46). This way, we don't need to duplicate every login and connect method, and it remains optional.

So my idea is:

* Revert the addition of `bot_sign_in_with_proxy`, `request_login_code_with_proxy`, and `connect_with_proxy` in the `client`.
* Remove `proxy: Option<String>` from `connect_sender` in the `client`.

I'm okay with keeping the `proxy` parameter in the `mtsender` crate for the time being (that's not really public API I care much about for the time being).

We can also merge the removal of the generic `A: ToSocketAddrs` for the time being, but I'd like to open an issue after we merge so we don't forget to fix that. This does not need to be fixed in this PR.

**Comment by niuhuan** [comment]
I' had modify `SocksAddr` to `ToSocksAddr + IntoTargetAddr`, It looks good now.

**Comment by niuhuan** [comment]
DEPS.md has been update

File: lib/grammers-client/src/client/client.rs
Comment (Lonami) [inline lib/grammers-client/src/client/client.rs:95]:
```suggestion
    /// URL of the proxy to use.
    ///
    /// The scheme must be `socks5`. Username and password are optional.
    ///
    /// Both a host and port must be provided. If a domain is used for the host, domain, its address will be looked up.
```

File: lib/grammers-mtsender/DEPS.md
Comment (Lonami) [inline lib/grammers-mtsender/DEPS.md:47]:
```suggestion
## url

Used to parse the optional proxy URL.

## trust-dns-resolver

Used to look up the IP address of the proxy host if a domain is provided.

## tokio-socks

SOCKS5 proxy support.

## rand

Used to choose one of the many IP addresses of the proxy's domain.
```

File: lib/grammers-client/DEPS.md
Comment (Lonami) [inline lib/grammers-client/DEPS.md:97]: `grammers-client` does not need to be updated because its `Cargo.toml` does not depend on these so we don't need to update this file.

```suggestion
```

**Review by Lonami — CHANGES_REQUESTED** [review]

**Review by Lonami — APPROVED** [review]

**Comment by Lonami** [comment]
Thanks. I still plan to make further changes but we can merge it for now.

**Comment by Lonami** [comment]
Hi @niuhuan, would you mind testing the above commit? Notably:

* I have made the proxy feature optional. You must enable it in order to specify a proxy.
* I have removed the dependency on `rand`. I don't think randomness is a good fit for this task.
* I have renamed the field to `proxy_url` because I think that's clearer for the end user.

I've also changed a bit how the code is structured. It should still work fine, but I would like it if you could confirm if these changes work for you.

**Comment by niuhuan** [comment]
That's good, I had test on my windows desktop, It's running. Thanks

**Pull request merged**
